### ISSUE TITLE
Misleading error when return type `overwrites` type inside function

### ISSUE THREAD

**Issue by WiseBird** [Microsoft/TypeScript#25644]
**TypeScript Version:**  3.0.0-dev.201xxxxx

**Search Terms:** type inference return type infers

**Code**

```ts
function pipe<A, B>(obj: A, op1: (source: A) => B): B {
    return null;
}
function tap<T>(next: (x: T) => void): (source: T) => T {
    return null;
}

function func(): Object {
    return pipe(10, tap(x => {
        console.log(x.toExponential);
    }));
}
```

**Expected behavior:**

TS should either infers `number` as `x`'s type or show error on return type as it really hard to understand at first what the cause.

**Actual behavior:** TS complains that `Property 'toExponential' does not exist on type 'Object'.`

**Playground Link:** [Link](https://www.typescriptlang.org/play/index.html#src=function%20pipe%3CA%2C%20B%3E(obj%3A%20A%2C%20op1%3A%20(source%3A%20A)%20%3D%3E%20B)%3A%20B%20%7B%0D%0A%20%20%20%20return%20null%3B%0D%0A%7D%0D%0Afunction%20tap%3CT%3E(next%3A%20(x%3A%20T)%20%3D%3E%20void)%3A%20(source%3A%20T)%20%3D%3E%20T%20%7B%0D%0A%20%20%20%20return%20null%3B%0D%0A%7D%0D%0A%0D%0Afunction%20func()%3A%20Object%20%7B%0D%0A%20%20%20%20return%20pipe(10%2C%20tap(x%20%3D%3E%20%7B%0D%0A%20%20%20%20%20%20%20%20console.log(x.toExponential)%3B%0D%0A%20%20%20%20%7D))%3B%0D%0A%7D)

**Comment by mhegazy** [comment]
There is a return type inference that is happening from `Object`. since the compiler knows you are assigning the result of `pipe` call to `Object`, then it infers `Object` for `B`, and then propagates that to `tap`, and infers that to `T`. `T` happens to have another inference which is `number` from `A`'s inference in `pipe`, but `Object` happens to be a supper type of `number`, so the inference picks the most general type `Object`.
